software/contrib/egressus_melodiam.py

```diff
@@ -0,0 +1,910 @@
+There are both linear and non-linear interpolation functions that create various smooth and no-so-smooth shapes.
+
+Each slew shape function generates an array of values (samples) between two given points.
+Samples are output on the CV outputs based on the list of pre-computed interpolated values (sample buffers)
+A value from the sample buffer is retrieved using a Python generator function (self.slewGenerator()).
+One instance of the generator function for each each slew shape exists in a list: self.slewGeneratorObjects[]
+
+A new sample buffer (array of interpolated values) is created at each clock step.
+The number of samples required in each sample buffer (one for each output) is calculated at each clock step or if
+the clock rate or output division changes.
+
+The timing and associated hanlding of input clock interrupts (clocks to din) is not perfect - python/micropython is not
+known to be great at sub-second timing!
+Therefore, at the end of each clock step the algorithm tunes itself to work around sample buffer overruns or underruns.
+If there is a buffer underrun (not enough samples in the buffer), the previous output voltage (sample) is used until the algorithm
+catches back up with itself.
+
+slewGeneratorObjects[] contains 6 object (one for each output)
+Each slewGeneratorObjects[] object is a reference to a copy of the self.slewGenerator() function.
+The self.slewGenerator() function receives a buffer filled with samples and yields one sample each time it is called using next()
+
+In order to maintain the best balance of smooth waves and Rpi pico memory usage, an algorithm is used to vary the 
+sample rate automatically based on the selected output division.
+This causes the sample rate to be at minium when there is a slow clock and high output division.
+Conversely, when there is a fast clock and low output division a higher sample rate is used to avoid unwanted steps
+often caused by under-samping a wave form.
+
+"""
+# Minimum allowed time between incoming 16th note clocks. Smaller values are too fast for poor ole micropython to keep up
+MIN_CLOCK_TIME_MS = 50
+
+# Maximum allowed time between incoming 16th note clocks. This is 2 BPM, larger value causes memory issues unless the next two vals are halfed
+MAX_CLOCK_TIME_MS = 3750
+
+# Max sample rate and clock divisor to keep required buffer sizes within memory limits
+MAX_SAMPLE_RATE = 32
+MAX_OUTPUT_DENOMINATOR = 8
+
+# Calculate maximum sample buffer size required
+SLEW_BUFFER_SIZE_IN_SAMPLES = int(
+    (MAX_CLOCK_TIME_MS / 1000) * MAX_SAMPLE_RATE * MAX_OUTPUT_DENOMINATOR
+)
+
+# Reduce knob hysteresis using this value - Mutable Instruments style
+KNOB_CHANGE_TOLERANCE = 0.999
+
+# Set the maximum CV voltage using a global config value
+# Important: Needs firmware v0.12.1 or higher
+MAX_CV_VOLTAGE = europi_config["max_output_voltage"]
+
+# Set the desired CPU frequency using a global config value. Default is overclocked.
+CPU_FREQ = europi_config["cpu_freq"]

```

Comment:
[required]
This can be removed; it's not necessary unless you're doing any math that relies on the clock frequency (which I don't think you are).

Reply:
done. I didn't realize the firmware did this already